src/indexing.jl

```diff
@@ -0,0 +1,399 @@
+using Base: @_propagate_inbounds_meta
+
+# Indexing interface
+# ------------------
+# these definitions are not using @derive since we need the @inline annotation
+# to correctly deal with boundschecks and @inbounds
+
+"""
+    getstoredindex(A::AbstractArray, I...) -> eltype(A)
+
+Obtain `getindex(A, I...)` with the guarantee that there is a stored entry at that location.
+
+Similar to `Base.getindex`, new definitions should be in line with `IndexStyle(A)`.
+"""
+@inline getstoredindex(A::AbstractArray, I...) =
+  @interface interface(A) getstoredindex(A, I...)
+
+"""
+    getunstoredindex(A::AbstractArray, I...) -> eltype(A)
+
+Obtain `getindex(A, I...)` with the guarantee that there is no stored entry at that location.
```

Comment:
Is there a guarantee of that? Also, the way I was thinking about this function was that it outputs what _would be_ the unstored value at the specified index, whether or not there is a stored value there in practice. But I'm open to suggestions about that, maybe it should have a different name.